### DISCUSSION TITLE
History_meta 1.4 to 2.0 migration - How to exclude specified columns?

### DISCUSSION THREAD

**Discussion by mmssix** [sqlalchemy/sqlalchemy#10618] [Usage Questions]
I've googled, and searched the discussions here, but didn't see anything applicable to my issue.  Hoping someone might have a minute to assist.  Thanks for looking.

I'm trying to mod the new 2.0 history_meta to exclude certain specified columns at creation time.  I had this working in 1.4 with the code below.  I don't know if this was the right way, but it did function well.

The issue is that I'm not a good enough programmer to figure out how to do it with this new 2.0 code.  The 1.4 method below, used on the new code, still creates the excluded columns.  I tried using `remove()` on the `local_mapper` object at the top of `_history_mapper`, but that causes other issues and and does not do the job.

1.4 Example:
```python
to_ignore = {'updated_at', 'created_at'}

def _is_versioning_col(col):
    if col.key in to_ignore:
        return True
    return "version_meta" in col.info
```
<br>
I also tried removing the columns after `versioned_cls` was created (at the very bottom of `_history_mapper`) with something like this.  It worked for writes, but pulling the history causes odd sqlalchemy errors.

```python
table_meta = versioned_cls.metadata.tables[versioned_cls.__tablename__]
for each in [x.key for x in table_meta._columns if x.key in to_ignore]:
    table_meta._columns.remove(table_meta._columns[each])
```

<br><br>
To illustrate: Source model and the SQL for the desired history table
```python
# Seed table
class DemoTestTable(Versioned, db.Model):
    __tablename__ = "vtest"
    id = db.Column(db.Integer, primary_key=True)
    name = db.Column(db.String(254))
    created_at = db.Column(db.TIMESTAMP, server_default=text('CURRENT_TIMESTAMP'))
    updated_at = db.Column(db.TIMESTAMP, server_default=text('CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP'))
```
```sql
# Desired dynamically created table
CREATE TABLE `vtest_history` (
	`id` INT(11) NOT NULL,
	`name` VARCHAR(254) NULL DEFAULT NULL COLLATE 'utf8mb3_unicode_ci',
	`version` INT(11) NOT NULL,
	`changed` DATETIME NULL DEFAULT NULL,
	PRIMARY KEY (`id`, `version`) USING BTREE,
);
```

**Comment by zzzeek** [comment]
the old version copies columns individually here:  https://github.com/sqlalchemy/sqlalchemy/blob/f04d221b62c92ce4cc9e7ee374c3ca566fd35f94/examples/versioned_history/history_meta.py#L68  which makes use of "column.copy()" which was never meant to be public API so that's removed in 2.0.

the new version uses the more simple "to_metadata" approach here: https://github.com/sqlalchemy/sqlalchemy/blob/aae161f2200456366a692410e40f429c8648c526/examples/versioned_history/history_meta.py#L55 , this actually copies the columns for you, and greatly simplifies the whole recipe.

the Table, once created, does not have public API to remove columns, and there is not right now an API feature in `to_metadata()` that allows columns to be excluded.   So the `to_metadata()` feature here would need to be replaced with the old approach in order to exclude columns, which relies  on using very careful manual copying of columns as was done in the previous version, preferably without using the private "copy()" functions either.

im not really sure why you'd need to exclude columns from the table itself, they can just be set to "nullable" and ignored.   you can also affect how the mapping is created here: https://github.com/sqlalchemy/sqlalchemy/blob/aae161f2200456366a692410e40f429c8648c526/examples/versioned_history/history_meta.py#L177  adding the [exclude_properties](https://docs.sqlalchemy.org/en/20/orm/mapping_api.html#sqlalchemy.orm.Mapper.params.exclude_properties) mapper parameter to exclude those columns that you want to ignore.

**Reply by mmssix** [reply]
So its not possible to remove the columns from the local_mapper metadata above `history_table = local_mapper.local_table.to_metadata(` so they never get added to the new history model in the first place?

I did try adding the "exclude_properties", but the columns were still added to the new history table.
```python
    versioned_cls = type(
        "%sHistory" % cls.__name__,
        bases,
        {
            "_history_mapper_configured": True,
            "__table__": history_table,
            "__mapper_args__": dict(
                exclude_properties=['updated_at', 'created_at'],
                inherits=super_history_mapper,
                polymorphic_identity=local_mapper.polymorphic_identity,
                polymorphic_on=polymorphic_on,
                properties=properties,
            ),
        },
    )
```

My use case is that those columns are part of of an AbstractBase class I apply to every table.  Storing those values for 10's or 100's of thousands of rows just takes up space.  Its the other columns I care about the changes in.  The updated_at and created_at are redundant for the purposes of the history entries, but crucial to the function of the parent table.

**Comment by zzzeek** [comment]
> So its not possible to remove the columns from the local_mapper metadata above `history_table = local_mapper.local_table.to_metadata(` so they never get added to the new history model in the first place?

the to_metadata() method makes a copy of a table with an optional new name.  it does not currently have the option to exclude specific columns (though there is no technical reason this could not be a new feature).

Once you have a Table, then you map it.  the "exclude_properties" omits them *from the mapping*, but not from the table.

My suggestion is that unless you want to revert part of the recipe to the 1.4 version that does the "to_metadata" part more manually, you just allow these columns to be in the table and they remain NULL.  they take up no room in your database if they are not populated.

Otherwise, go into the source code here:  https://github.com/sqlalchemy/sqlalchemy/blob/fc6c2b19fd7f5cb89e0c405f5aa3b3360e4c4a93/lib/sqlalchemy/sql/schema.py#L1444 and just vendor (i.e. copy it to your own source code) this method for now, adding the "exclude_columns" argument to the method.  that could be a future pull request for SQLAlchemy.

> My use case is that those columns are part of of an AbstractBase class I apply to every table. Storing those values for 10's or 100's of thousands of rows just takes up space. Its the other columns I care about the changes in. The updated_at and created_at are redundant for the purposes of the history entries, but crucial to the function of the parent table.

if you exclude these columns from the mapping and ensure they have no server side defaults, they remain empty.   otherwise look to vendor the to_metadata() method to write a table copy routine that suits your needs.

**Reply by CaselIT** [reply]
> the to_metadata() method makes a copy of a table with an optional new name. it does not currently have the option to exclude specific columns (though there is no technical reason this could not be a new feature).

I've created https://github.com/sqlalchemy/sqlalchemy/issues/10622 to track it

**Reply by mmssix** [reply]
I'll give these alternate methods a try tonight.  If they dont work, I'll just live with the columns being there.

Thank you so much for your time.  The open source community really is amazing.  I wish I had more to offer it than the occasional noob question.
